src/d3a/models/strategy/__init__.py

```diff
@@ -454,14 +455,28 @@ def __init__(self):
         self._bids = {}
         self._traded_bids = {}
 
-    def post_bid(self, market, price, energy, buyer_origin=None):
+    def _remove_existing_bids(self, market: Market) -> None:
+        """Remove all existing bids in the market"""
+
+        for bid in self.get_posted_bids(market):
+            assert bid.buyer == self.owner.name
+            if bid.id in market.bids.keys():
+                bid = market.bids[bid.id]
+
+            self.remove_bid_from_pending(market.id, bid.id)
+
+    def post_bid(self, market, price, energy, replace_existing=False, buyer_origin=None):
+        if replace_existing:
+            self._remove_existing_bids(market)
+
         bid = market.bid(
             price,
             energy,
             self.owner.name,
             original_bid_price=price,
-            buyer_origin=buyer_origin
-        )
+            buyer_origin=buyer_origin,
+            replace_existing=replace_existing)
```

Comment:
I dont think this is needed and could also lead to confusion.
IMO you "only" need to do the following when the `replace_existing` flag was set via the api_client:
1. remove all bids from the BaseStrategy tracker and collect their uuids
2. use these uuids and delete them from the market (see how it is implemented in delete_bid).
Let me know if you need more information. Thanks!

Reply:
I agree that this is redundant and not that optimal.
However, we want to find a way to add the value of `replace_existing` inside the `Bid` object, because that's what will be returned to the client. And the `Bid` object is created inside of `market.bid()` at the moment:
https://github.com/gridsingularity/d3a/pull/1038/files/9143d4cc7a88278692e1fe523fbc973967098cf5#diff-e048ea2bf894f98b35d1c8798a95454a8dd7507f5b1d03e592d8ad65345c85b2R81-R85

Unfortunately, I couldn't come up with a cleaner idea to do that yet.

Reply:
This is probably a design question and that is why I would like to consult @spyrostz here.
However IMO, the market (and neither the Blockchain) does not care about whether the bid is a "replace_existing" one and frankly this could lead to some privacy issues too, because this way others can see your bidding strategy very easily.
I still dont get why you want to track it. What does the client need to know? IMO only that the new bid was places and that the old ones where deleted. If useful, you can even return the uuids of the deleted bids to the client.
What do you think?

Reply:
IMO it is far better to manually append it on the `self.redis.publish_json` call, than adding it to the Bid namedtuple. In general, d3a (the exchange itself) is not interested in this flag, the API user that sent this command might be interested in it though. 
For instance, for the bid case, the response is transmitted via this call:
`            self.redis.publish_json(
                bid_response_channel,
                {"command": "bid", "status": "ready", "bid": bid.to_JSON_string(),
                 "transaction_id": arguments.get("transaction_id", None)})`

The actual response of the command is this: `"bid": bid.to_JSON_string()`, therefore you could add one more key-value pair to the bid JSON to reflect the value of `replace_existing`. Hope these all make sense.

Reply:
@hannesdiedrich I hadn't thought about the privacy concerns. That's extremely interesting, thanks!

@spyrostz Just to be sure: you are suggesting to add it to `bid.to_JSON_string()`, and not as a separate key-value to the response (like `transaction_id`). Is this correct?

Reply:
@fievelk yes, I am suggesting to add it to bid.to_JSON_string(), but as an optional method argument (for instance, by converting the method to def to_JSON_string(replace_existing=None). In this case you can pass the replace_existing variable and populate the bid json string with its value if the replace_existing argument is used, but if it is not used you can entirely omit it from the json string)